streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java

```diff
@@ -244,4 +238,31 @@ public void shouldLogAndMeterWhenSkippingNullKey() {
             assertThat(appender.getMessages(), hasItem("Skipping record due to null key. value=[1] topic=[topic] partition=[0] offset=[0]"));
         }
     }
+
+    @Test
+    public void shouldLogAndMeterWhenSkippingExpiredWindow() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        final String topic = "topic";
+
+        final KStream<String, String> stream1 = builder.stream(topic, Consumed.with(Serdes.String(), Serdes.String()));
+        stream1.groupByKey(Serialized.with(Serdes.String(), Serdes.String()))
+            .windowedBy(TimeWindows.of(10).advanceBy(5).until(100))
+            .aggregate(
+                MockInitializer.STRING_INIT,
+                MockAggregator.toStringInstance("+"),
+                Materialized.<String, String, WindowStore<Bytes, byte[]>>as("topic1-Canonicalized").withValueSerde(Serdes.String())
+            );
+
+        final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
+        try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props, 0L)) {
+            driver.pipeInput(recordFactory.create(topic, "k", "2", 101L));
+            driver.pipeInput(recordFactory.create(topic, "k", "3", 0L));
+            LogCaptureAppender.unregister(appender);
+
+            assertEquals(1.0, getMetricByName(driver.metrics(), "skipped-records-total", "stream-metrics").metricValue());
+            assertThat(appender.getMessages(), hasItem(
+                "Skipping record for expired window. key=[k] topic=[topic] partition=[0] offset=[1] timestamp=[0] window=[0] expiration=[1]"
```

Comment:
@mjsax @bbejeck @guozhangwang , Can you all double-check my math here. Is this the correct time for window zero to expire: at stream-time 101 instead of stream time 100 ?

Reply:
I think it should expire at 100 already, because, the window is `[0,10)` (=`[0,9]`). We always required `until >= size`. For example, with size=10 and until=10 no late data can be handled: a record for the window at stream-time=10 would be dropped already. Thus, with `until=100`, a record should be dropped at stream-time=100. (From my understanding -- feel free to correct me). The "issue" is, that `until` is _not_ window-end-time plus delay... atm -- and if it would be, we would specify the delay as 90, what means that a record must be processed within 9+90, and is lated at 10+90.

Should we maybe test with late records ts=9 and ts=10, too? To test the edge case? 9 should be dropped, while 10 should be processed.

Reply:
Btw: I think we should also check the produced output.

Reply:
Thanks @mjsax , I kept going in circles over what the right boundaries are.

Reply:
+1. BTW for stream-stream windowed join, the join windows should be actually [start, end], i.e. both inclusive.